cloud/linode/cilium_loadbalancers.go

```diff
@@ -0,0 +1,355 @@
+package linode
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"net/http"
+	"strings"
+
+	"github.com/cilium/cilium/pkg/k8s/apis/cilium.io/v2alpha1"
+	ciliumclient "github.com/cilium/cilium/pkg/k8s/client/clientset/versioned/typed/cilium.io/v2alpha1"
+	slimv1 "github.com/cilium/cilium/pkg/k8s/slim/k8s/apis/meta/v1"
+	"github.com/google/uuid"
+	"github.com/linode/linodego"
+	v1 "k8s.io/api/core/v1"
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/clientcmd"
+	"k8s.io/klog/v2"
+)
+
+const (
+	ciliumLBClass              = "io.cilium/bgp-control-plane"
+	ipHolderLabel              = "linode-ccm-ip-holder"
+	ciliumBGPPeeringPolicyName = "linode-ccm-bgp-peering"
+)
+
+// This mapping is unfortunately necessary since there is no way to get the
+// numeric ID for a data center from the API.
+// These values come from https://www.linode.com/docs/products/compute/compute-instances/guides/failover/#ip-sharing-availability
+var (
+	regionIDMap = map[string]int{
+		"us-southeast": 4,  // Atlanta, GA (USA)
+		"us-ord":       18, // Chicago, IL (USA)
+		"us-central":   2,  // Dallas, TX (USA)
+		// "us-west"   : 3,  // Fremont, CA (USA) UNDERGOING NETWORK UPGRADES
+		"us-lax":     30, // Los Angeles, CA (USA)
+		"us-mia":     28, // Miami, FL (USA)
+		"us-east":    6,  // Newark, NJ (USA)
+		"us-sea":     20, // Seattle, WA (USA)
+		"us-iad":     17, // Washington, DC (USA)
+		"ca-central": 15, // Toronto (Canada)
+		"br-gru":     21, // São Paulo (Brazil)
+		// EMEA
+		"nl-ams":     22, // Amsterdam (Netherlands)
+		"eu-central": 10, // Frankfurt (Germany)
+		"eu-west":    7,  // London (United Kingdom)
+		"it-mil":     27, // Milan (Italy)
+		"ap-west":    14, // Mumbai (India)
+		"fr-par":     19, // Paris (France)
+		"se-sto":     23, // Stockholm (Sweden)
+		// APAC
+		"in-maa":       25, // Chennai (India)
+		"id-cgk":       29, // Jakarta (Indonesia)
+		"jp-osa":       26, // Osaka (Japan)
+		"ap-south":     9,  // Singapore
+		"ap-southeast": 16, // Sydney (Australia)
+		"ap-northeast": 11, // Tokyo (Japan)
+	}
+	errNoBGPSelector = errors.New("no BGP node selector set to configure IP sharing")
+)
+
+// createSharedIP requests an additional IP that can be shared on Nodes to support
+// loadbalancing via Cilium LB IPAM + BGP Control Plane.
+func (l *loadbalancers) createSharedIP(ctx context.Context, nodes []*v1.Node) (string, error) {
+	if Options.BGPNodeSelector == "" {
+		return "", errNoBGPSelector
+	}
+
+	ipHolder, err := l.ensureIPHolder(ctx)
+	if err != nil {
+		return "", err
+	}
+
+	newSharedIP, err := l.client.AddInstanceIPAddress(ctx, ipHolder.ID, true)
+	if err != nil {
+		return "", err
+	}
+
+	// need to retrieve existing public IPs on the IP holder since ShareIPAddresses
```

Comment:
I think instead here so we can have a single "ip-holder" for multiple clusters, I should just gather up the IPs from the existing (if any) CiliumLoadBalancerPools created by the CCM to determine the full list of IPs to share on the selected nodes